File: branches/sdk-0-5/implementation/UrakawaToolkit/IChannelFactory.cs
using System;

namespace urakawa.core
{
	/// <summary>
	/// Interface for an abstract factory creating <see cref="IChannel"/>s
	/// </summary>
	public interface IChannelFactory
	{
    /// <summary>
    /// Creates a new Channel with a given name, which is not linked to the channels list yet. 
    /// </summary>
    /// <param name="name">The name of the new <see cref="Channel"/></param>
    /// <returns>The new <see cref="Channel"/></returns>
    /// <exception cref="exception.MethodParameterIsNullException">
    /// Thrown when <paramref name="name"/> is null
    /// </exception>
    /// <exception cref="exception.MethodParameterIsEmptyStringException">
    /// Thrown when <paramref name="name"/> is an empty string
    /// </exception>
		IChannel createChannel(string name);

		/// <summary>
		/// Creates a new <see cref="IChannel"/> matching a given QName
		/// </summary>
		/// <param name="localName">The local part of the QName</param>
		/// <param name="namespaceUri">The namespace uri part of the QName</param>
		/// <returns>The created <see cref="IChannel"/> or <c>null</c> is the given QName is not supported</returns>
		IChannel createChannel(string localName, string namespaceUri);
	}
}

File: java/core/src/org/daisy/urakawa/property/channel/IChannelFactory.java
package org.daisy.urakawa.property.channel;

import org.daisy.urakawa.IPresentation;
import org.daisy.urakawa.IWithPresentation;
import org.daisy.urakawa.exception.IsNotInitializedException;
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.xuk.IXukAble;

/**
 * <p>
 * This is the factory that creates
 * {@link org.daisy.urakawa.property.channel.IChannel} instances.
 * </p>
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Create - org.daisy.urakawa.property.channel.IChannel
 * @depend - Aggregation 1 org.daisy.urakawa.IPresentation
 */
public interface IChannelFactory extends IXukAble, IWithPresentation {
	/**
	 * Convenience method that delegates to the IPresentation
	 * 
	 * @see IPresentation#getChannelsManager()
	 * @return the IChannelsManager
	 * @throws IsNotInitializedException 
	 */
	public IChannelsManager getChannelsManager()
			throws IsNotInitializedException;

	/**
	 * <p>
	 * Creates a default channel.
	 * </p>
	 * <p>
	 * The returned object is managed by its associated manager.
	 * </p>
	 * 
	 * @return cannot return null
	 */
	public IChannel createChannel();

	/**
	 * <p>
	 * Creates a channel.
	 * </p>
	 * <p>
	 * The returned object is managed by its associated manager.
	 * </p>
	 * <p>
	 * This factory method takes arguments to specify the exact type of object
	 * to create, given by the unique QName (XML Qualified Name) used in the XUK
	 * serialization format. This method can be used to generate instances of
	 * subclasses of the base object type.
	 * </p>
	 * 
	 * @param xukLocalName
	 *            cannot be null, cannot be empty string.
	 * @param xukNamespaceURI
	 *            cannot be null, but can be empty string.
	 * @return can return null (in case the QName specification does not match
	 *         any supported type).
	 * @tagvalue Exceptions "MethodParameterIsNull-MethodParameterIsEmptyString"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 * @throws MethodParameterIsEmptyStringException
	 *             Empty string '' method parameter is forbidden:
	 *             <b>xukLocalName</b>
	 */
	public IChannel createChannel(String xukLocalName, String xukNamespaceURI)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException;
}
